torchtune/utils/memory.py

```diff
@@ -6,29 +6,131 @@
 
 import gc
 
-from typing import Any, Dict, Optional, Set
+from collections import defaultdict
```

Comment:
Changes in this file are a bit dense and I'm parsing them right now, but one thought I had was can we keep the current functionality and just add this as an additive change for now while we thoroughly stress test this? We can just update our default configs once we have tested this across fine-tunes and hardware types. WDYT?

Reply:
sure we can do that.  Do you have a way you want to toggle between old and new in the config or else I can just add a flag for use old or new as a temp solution.

Reply:
Good question, maybe before I answer that another noob question. I noticed the current ac changes go before the FSDP wrapping and the previous one after - does that matter?

Reply:
Actually, sorry for the confusion on this. This particular function is used across many recipes. So this is my concrete proposal:

- can you move this functionality to maybe activation.py or something similar? I'd mention in the docstrings that this is under development or something like that and so we dont need to worry about BC or anything like that
- add a new config under configs/dev/7B_full.yaml with these changes

over time we can move this dev config to be the default and then also deprecate the stuff from memory.py. WDYT?

Reply:
sure, activation.py sounds like a very modular way to implement this.
will work on that now and make a new config of 7B_full_experimental.yaml.

Reply:
fyi, I've moved things into a new file activations.py and also added a new 7B_full_experimental.yaml file.
I then registered that yaml file in the tune registry as distributed recipe and everything seems to be working.
Internally, I've modified the yaml and code to be forward and back compat.
Specifically:
yaml file keeps the enable_activation_checkpointing: bool for old and new. 
True = do ac
False = do not do ac for all.
From there we check if ac_mode is present in the yaml - if not, then it just goes to the old AC assuming enable is true.
Otherwise it goes to the new activations.py and based on ac_mode ('full', 'selective') and option(int or 'op') does the right selective or full ac impl.